### PR TITLE
Add native input validation 1.2

### PR DESCRIPTION
**Pull request by TerentievEvgany**

### CODE REVIEW

File: src/main/java/org/prebid/server/validation/model/ContextSubType.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package org.prebid.server.validation.model;
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/model/ContextSubType.java]: Since these enums are related to OpenRTB Native in general than to validation logic I think it's logical to move them into com.iab.openrtb.request.native package

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 2]
```diff
@@ -365,17 +374,64 @@ private static Request parseNativeRequest(String rawStringNativeRequest, int imp
         }
     }
 
-    private void validateNativeContext(Integer context, int index) throws ValidationException {
-        if (context != null && (context < 1 || context > 3)) {
+    private void validateNativeContextTypes(Integer context, Integer contextSubType, int index)
+            throws ValidationException {
+        if (context != null && (context < ContextType.CONTENT.getValue() || context > ContextType.PRODUCT.getValue())) {
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Implicit unboxing here and in the code that follows is a bit of concern to me. Unnecessary unboxing could be avoided if `context` and `contextSubType` values were saved in `int` local variables first and those local variables were used in following checks. It seems that `0` is safe to use if corresponding `Integer` is `null`.

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 3]
```diff
@@ -508,13 +564,50 @@ private void validateNativeVideoProtocols(List<Integer> protocols, int impIndex,
...
+        if (CollectionUtils.isNotEmpty(eventTrackers)) {
+            for (int eventTrackerIndex = 0; eventTrackerIndex < eventTrackers.size(); eventTrackerIndex++) {
+                validateNativeEventTracker(eventTrackers.get(eventTrackerIndex), impIndex, eventTrackerIndex);
+            }
+        }
+    }
+
+    private void validateNativeEventTracker(EventTracker eventTracker, int impIndex, int eventIndex)
+            throws ValidationException {
+        if (eventTracker != null && (eventTracker.getEvent() < EventType.IMPRESSION.getValue()
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Please extract `eventTracker.getEvent()` result to a local variable and mind possible `null`

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 4]
```diff
@@ -508,13 +564,50 @@ private void validateNativeVideoProtocols(List<Integer> protocols, int impIndex,
...
+                || eventTracker.getEvent() > EventType.VIEWABLE_VIDEO50.getValue())) {
+            throw new ValidationException(
+                    "request.imp[%d].native.request.eventtrackers[%d].event is invalid. See section 7.6: https://iabtechlab.com/wp-content/uploads/2016/07/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=43", impIndex, eventIndex
+            );
+        }
+
+        List<Integer> methods = eventTracker.getMethods();
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Please make the variable final

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 5]
```diff
@@ -508,13 +564,50 @@ private void validateNativeVideoProtocols(List<Integer> protocols, int impIndex,
...
+
+        if (CollectionUtils.isEmpty(methods)) {
+            throw new ValidationException(
+                    "request.imp[%d].native.request.eventtrackers[%d].method is required. See section 7.7: https://iabtechlab.com/wp-content/uploads/2016/07/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=43", impIndex, eventIndex
+            );
+        }
+
+        for (int methodIndex = 0; methodIndex < methods.size(); methodIndex++) {
+            if (methods.get(methodIndex) < EventTrackingMethod.IMAGE.getValue()
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Please extract `methods.get(methodIndex)` result to a local variable and mind possible `null`

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 6]
```diff
@@ -365,17 +374,64 @@ private static Request parseNativeRequest(String rawStringNativeRequest, int imp
...
     private void validateNativePlacementType(Integer placementType, int index) throws ValidationException {
-        if (placementType != null && (placementType < 1 || placementType > 4)) {
+        if (placementType != null && (placementType < PlacementType.FEED.getValue()
+                || placementType > PlacementType.RECOMMENDATION_WIDGET.getValue())) {
             throw new ValidationException(
-                    "request.imp[%d].native.request.plcmttype must be in the range [1, 4]. Got %d",
+                    "request.imp[%d].native.request.plcmttype is invalid. See https://iabtechlab.com/wp-content/uploads/2016/07/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=40",
                     index, placementType);
         }
     }
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Please address removal of ending `.` in several error messages in this file.

File: src/test/java/org/prebid/server/validation/RequestValidatorTest.java [hunk 7]
```diff
@@ -1279,7 +1280,113 @@ public void validateShouldReturnValidationResultWithErrorWhenContentTypeOutOfPos
 
         // then
         assertThat(result.getErrors()).hasSize(1)
-                .containsOnly("request.imp[0].native.request.context must be in the range [1, 3]. Got 100");
+                .containsOnly("request.imp[0].native.request.context is invalid. See https://iabtechlab.com/wp-content/uploads/2016/07/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=39");
+    }
+
+    @Test
+    public void validateShouldReturnValidationResultWithErrorWhenContentSubTypeOutOfPossibleValuesRange()
```
Comment (schernysh) [inline src/test/java/org/prebid/server/validation/RequestValidatorTest.java]: Please add tests that cover other possible invalid contexType/contextSubtypes combinations and correct handling of null values as well.

Comment (schernysh) [inline src/test/java/org/prebid/server/validation/RequestValidatorTest.java]: Here and below method name should include `Context` instead of `Content`

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 8]
```diff
@@ -365,17 +374,72 @@ private static Request parseNativeRequest(String rawStringNativeRequest, int imp
...
 
-    private void validateNativeContext(Integer context, int index) throws ValidationException {
-        if (context != null && (context < 1 || context > 3)) {
-            throw new ValidationException(
-                    "request.imp[%d].native.request.context must be in the range [1, 3]. Got %d", index, context);
+    private void validateNativeContextTypes(Integer context, Integer contextSubType, int index)
+            throws ValidationException {
+
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Extra newline

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 9]
```diff
@@ -365,17 +374,72 @@ private static Request parseNativeRequest(String rawStringNativeRequest, int imp
...
+
+        if (context != null) {
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: I'm not sure that this `if` is needed, type could be initialized as follows `final int type = context != null ? context.intValue() : 0`. Same is true for `subType`

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 10]
```diff
@@ -508,13 +572,56 @@ private void validateNativeVideoProtocols(List<Integer> protocols, int impIndex,
...
+            for (int eventTrackerIndex = 0; eventTrackerIndex < eventTrackers.size(); eventTrackerIndex++) {
+                validateNativeEventTracker(eventTrackers.get(eventTrackerIndex), impIndex, eventTrackerIndex);
+            }
+        }
+    }
+
+    private void validateNativeEventTracker(EventTracker eventTracker, int impIndex, int eventIndex)
+            throws ValidationException {
+        if (eventTracker != null) {
+            final Integer event = eventTracker.getEvent();
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: Same issue with autounboxing here

File: src/main/java/org/prebid/server/validation/RequestValidator.java [hunk 11]
```diff
@@ -508,13 +572,56 @@ private void validateNativeVideoProtocols(List<Integer> protocols, int impIndex,
...
+            final List<Integer> methods = eventTracker.getMethods();
+
+            if (CollectionUtils.isEmpty(methods)) {
+                throw new ValidationException(
+                        "request.imp[%d].native.request.eventtrackers[%d].method is required. See section 7.7: https://iabtechlab.com/wp-content/uploads/2016/07/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=43", impIndex, eventIndex
+                );
+            }
+
+            for (int methodIndex = 0; methodIndex < methods.size(); methodIndex++) {
+                int method = methods.get(methodIndex);
```
Comment (schernysh) [inline src/main/java/org/prebid/server/validation/RequestValidator.java]: What if `methods.get(methodIndex)` returned `null`?

File: src/test/java/org/prebid/server/validation/RequestValidatorTest.java [hunk 12]
```diff
@@ -1279,7 +1280,169 @@ public void validateShouldReturnValidationResultWithErrorWhenContentTypeOutOfPos
...
+        final ValidationResult result = requestValidator.validate(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).hasSize(1)
+                .containsOnly("request.imp[0].native.request.contextsubtype is invalid. See https://iabtechlab.com/wp-content/uploads/2016/07/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=39");
+
+    }
+
+    @Test
+    public void validateShouldReturnValidationResultWithErrorWhenContentSubTypeAndContextTypeOutOfPossibleContentValuesRange()
```
Comment (schernysh) [inline src/test/java/org/prebid/server/validation/RequestValidatorTest.java]: Should be `Context` instead of `Content`. Please fix in other names as well
